src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -340,44 +341,58 @@ Transactor::checkSign (PreclaimContext const& ctx)
 NotTEC
 Transactor::checkSingleSign (PreclaimContext const& ctx)
 {
-    auto const id = ctx.tx.getAccountID(sfAccount);
-
-    auto const sle = ctx.view.read(
-        keylet::account(id));
-    auto const hasAuthKey     = sle->isFieldPresent (sfRegularKey);
-
-    // Consistency: Check signature & verify the transaction's signing
-    // public key is authorized for signing.
-    auto const spk = ctx.tx.getSigningPubKey();
-    if (!publicKeyType (makeSlice (spk)))
+    // Check that the value in the signing key slot is a public key.
+    auto const pkSigner = ctx.tx.getSigningPubKey();
+    if (!publicKeyType(makeSlice(pkSigner)))
     {
         JLOG(ctx.j.trace()) <<
             "checkSingleSign: signing public key type is unknown";
         return tefBAD_AUTH; // FIXME: should be better error!
     }
 
-    auto const pkAccount = calcAccountID (
-        PublicKey (makeSlice (spk)));
+    // Look up the account.
+    auto const idSigner = calcAccountID(PublicKey(makeSlice(pkSigner)));
+    auto const idAccount = ctx.tx.getAccountID(sfAccount);
+    auto const sleAccount = ctx.view.read(keylet::account(idAccount));
+    bool const isMasterDisabled = sleAccount->isFlag(lsfDisableMaster);
+
+    if (ctx.view.rules().enabled(fixDisabledRegularKey))
+    {
+
+        if ((*sleAccount)[~sfRegularKey] == idSigner
+                || (!isMasterDisabled && idAccount == idSigner))
```

Comment:
Keeping in mind that this is the new flow, I believe that we are making the logic a little complicated to follow. I would prefer to see this block take the following shape:

```
if (ctx.view.rules().enabled(fixDisabledRegularKey))
{
    // Signature made using the regular key: accepted
    if ((*sleAccount)[~sfRegularKey] == idSigner)
        return tesSUCCESS;

    // Signature made using the master key: accepted if not disabled
    if (!sleAccount->isFlag(lsfDisableMaster) && idAccount == idSigner)
        return tesSUCCESS;

    // Signature made with a key that's not authorized to sign
    return tefBAD_AUTH;
}
```

I understand that this "removes" a theoretically useful error message but I highly doubt that the distinct error message is useful _in practice_. Whether the failure was because the master key is disabled or because the wrong key was used doesn't seem particularly actionable. I _can_ imagine a scenario (a client that knows both the master and the regular key tries to sign with the master key first and, if disabled, transparently resigns with the regular key) but honestly, I find that flow broken.

The cost of having the extra result is, to be sure, minor but it also complicates (if only slightly) one of the most critical parts of the code: checking whether a given public key is authorized to sign for a particular account. In that context, every single instruction matters. Every single additional `if` is an opportunity to get it wrong or mess something up.

Reply:
I'll say that I do find @nbougalis's reformulation significantly easier to read and reason about, especially with the comments above each condition.  I wonder is we can keep both this readability and the `tefMASTER_DISABLED`?  Consider something like...
```
if (ctx.view.rules().enabled(fixDisabledRegularKey))
{
    // Signature made using the regular key: accepted
    if ((*sleAccount)[~sfRegularKey] == idSigner)
        return tesSUCCESS;

    // Signature made using the master key: accepted if not disabled
    if (!isMasterDisabled && idAccount == idSigner)
        return tesSUCCESS;

    // Signature made using disabled master key
    if (isMasterDisabled && idAccount == idSigner)
        return tefMASTER_DISABLED;

    // Signature made with a key that's not authorized to sign
    return tefBAD_AUTH;
}
```

Reply:
I like the readability that comes with splitting up the success checks (especially because they can be commented separately), but I don't necessarily think we need to remove the other error message. For example,
```
if (ctx.view.rules().enabled(fixDisabledRegularKey))
{
    // Signature made using the regular key: accepted
    if ((*sleAccount)[~sfRegularKey] == idSigner)
        return tesSUCCESS;

    // Signature made using the master key: accepted if not disabled
    if (!isMasterDisabled && idAccount == idSigner)
        return tesSUCCESS;

    // Signature made using the master key: reject if disabled
    if (isMasterDisabled && idAccount == idSigner)
        return tefMASTER_DISABLED;

    // Signature made with a key that's not authorized to sign
    return tefBAD_AUTH;
}
```

Reply:
Thanks, @scottschurr, @ximinez, @thejohnfreeman.